lib/types.ts

```diff
@@ -253,23 +253,80 @@ export type DeepOmit<T extends DeepOmitModify<Filter>, Filter> = T extends Built
         {
           [K in Extract<RequiredKeys<T>, keyof Filter>]: Filter[K] extends true
             ? never
-            : T[K] extends DeepOmitModify<Filter[K]>
+            : T[K] extends DeepModify<Filter[K]>
             ? DeepOmit<T[K], Filter[K]>
             : T[K];
         },
         never
       >;
-type DeepOmitModify<T> =
+
+/** Recursively pick deep properties */
+export declare type DeepPick<T extends DeepModify<Filter>, Filter> = T extends Builtin
+  ? T
+  : T extends Map<infer KeyType, infer ValueType>
+  ? ValueType extends DeepModify<Filter>
+    ? Map<KeyType, DeepPick<ValueType, Filter>>
+    : T
+  : T extends ReadonlyMap<infer KeyType, infer ValueType>
+  ? ValueType extends DeepModify<Filter>
+    ? ReadonlyMap<KeyType, DeepPick<ValueType, Filter>>
+    : T
+  : T extends WeakMap<infer KeyType, infer ValueType>
+  ? ValueType extends DeepModify<Filter>
+    ? WeakMap<KeyType, DeepPick<ValueType, Filter>>
+    : T
+  : T extends Set<infer ItemType>
+  ? ItemType extends DeepModify<Filter>
+    ? Set<DeepPick<ItemType, Filter>>
+    : T
+  : T extends ReadonlySet<infer ItemType>
+  ? ItemType extends DeepModify<Filter>
+    ? ReadonlySet<DeepPick<ItemType, Filter>>
+    : T
+  : T extends WeakSet<infer ItemType>
+  ? ItemType extends DeepModify<Filter>
+    ? WeakSet<DeepPick<ItemType, Filter>>
+    : T
+  : T extends Array<infer ItemType>
+  ? ItemType extends DeepModify<Filter>
+    ? Array<DeepPick<ItemType, Filter>>
+    : T
+  : T extends Promise<infer ItemType>
+  ? ItemType extends DeepModify<Filter>
+    ? Promise<DeepPick<ItemType, Filter>>
+    : T
+  : OmitProperties<
+      {
+        [K in Extract<OptionalKeys<T>, keyof Filter>]+?: Filter[K] extends true
+          ? T[K]
+          : T[K] extends DeepModify<Filter[K]>
+          ? DeepPick<T[K], Filter[K]>
+          : never;
+      },
+      never
+    > &
+      OmitProperties<
+        {
+          [K in Extract<RequiredKeys<T>, keyof Filter>]: Filter[K] extends true
+            ? T[K]
+            : T[K] extends DeepModify<Filter[K]>
+            ? DeepPick<T[K], Filter[K]>
+            : never;
+        },
+        never
+      >;
+
+type DeepModify<T> =
```

Comment:
`DeepModify` should be named `Skeleton` as we want to have all possible combinations of keys without thinking about values

Can experiment with it when cover everything with tests and fix it

Reply:
Those tests that are failing in your TS Playground deal with _optional_ properties of ComplexNestedObject.
And yes, it doesn't work this way at the moment (I don't remember exactly why, but there was some reason, when I have time maybe I'll look into it).

Anyway, the DeepOmit type, that is already present in ts-essentials from long ago, works the same way, i.e. you cannot deep omit optional properties (you can try and test it). 
DeepPick feature was designed as complementary type for DeepOmit and as such it works just fine.
This means that you can release this feature, and we'll see later if something can be done about this particular case.

Reply:
> All those tests that are failing in your TS Playground deal with optional properties of ComplexNestedObject.
And yes, it doesn't work this way at the moment

Yeah, I know, I will try to fix it here, that's why I created another MR

> Anyway, the DeepOmit type, that is already present in ts-essentials from long ago, works the same way, i.e. you cannot deep omit optional properties (you can try and test it).

That's true, but as we're preparing for the release, I would improve it (as I also have time on the weekend) because anyway we touch it and change

> Then I'll make a new PR that fixes this behaviour in both DeepPick and DeepOmit.

You can work on top of this PR if it's fine for you
I didn't mean to change anything else except for that

Reply:
Done here, will keep `DeepModify` for now for only `DeepPick`

`DeepOmit` still uses `DeepOmitModify` to not escalate errors